### PR TITLE
Analytics: Add debugging to disallowed analytics

### PR DESCRIPTION
**Pull request by sirreal**
It's a real pain to test Google Analytics because it may be silently disabled in certain circumstances (config, donottrack, personally identifying information PII).

This PR wraps GA tracking functions to provide debug output when GA is disabled.
It also DRYs some shared logic.

Examples of difficult debugging: #20048 #19995

### Disallowed (dev env, donottrack, etc.)

disallowed

### Allowed (no change)

allowed

## Testing
1. `localStorage.debug = 'calypso:analytics:ga,calypso:analytics:utils'`.
1. Observe disallowed events (https://calypso.live/?branch=add/google-analytics-debug).
1. Run in allowed context, you'll need to test locally enabling via config (see patch below).
1. Observe allowed events (no change).

To enable on dev:

```patch
diff --git a/config/development.json b/config/development.json
index 07a63755dc..30a97d99cf 100644
--- a/config/development.json
+++ b/config/development.json
@@ -16,7 +16,7 @@
 	"boom_analytics_enabled": false,
 	"server_side_boom_analytics_enabled": false,
 	"boom_analytics_key": "development",
-	"google_analytics_enabled": false,
+	"google_analytics_enabled": true,
 	"google_analytics_key": "UA-10673494-15",
 	"google_adwords_conversion_id": 1067250390,
 	"google_adwords_conversion_id_jetpack": 937115306,
```

### CODE REVIEW

File: client/lib/analytics/index.js [hunk 1]
```diff
@@ -512,5 +507,30 @@ const analytics = {
...
+ *
+ * This method will display debug output if Google Analytics is suppresed, otherwise it will
+ * initialize and call the Google Analytics function it is passed.
+ *
+ * @see isGoogleAnalyticsAllowed)
+ *
+ * @param  {Function} func Google Analytics tracking function
+ * @return {Function}      Wrapped function
+ */
+function makeSafeDebuggableGoogleAnalyticsFunc( func ) {
```
Comment (sirreal) [inline client/lib/analytics/index.js]: Sorry, I really don't know what to call this…

Comment (tyxla) [reply client/lib/analytics/index.js]: It's not too bad. Maybe ditch `safe`?

File: client/lib/analytics/index.js [hunk 2]
```diff
@@ -512,5 +507,30 @@ const analytics = {
...
+ * initialize and call the Google Analytics function it is passed.
+ *
+ * @see isGoogleAnalyticsAllowed
+ *
+ * @param  {Function} func Google Analytics tracking function
+ * @return {Function}      Wrapped function
+ */
+function makeSafeDebuggableGoogleAnalyticsFunc( func ) {
+	return function( ...args ) {
+		if ( ! isGoogleAnalyticsAllowed() ) {
```
Comment (tyxla) [inline client/lib/analytics/index.js]: Looks like this one could use some tests?

File: client/lib/analytics/index.js (hunk 1 above)
Comment (sirreal) [reply client/lib/analytics/index.js]: Yeah, it initialized and protects agains disallowed calls. Maybe even just `makeGoogleAnalyticsTrackingFunction`?

Comment (tyxla) [reply client/lib/analytics/index.js]: Sounds better, go for it 👍

File: client/lib/analytics/index.js [hunk 3]
```diff
@@ -412,13 +412,10 @@ const analytics = {
...
-			if ( ! isGoogleAnalyticsAllowed() ) {
-				return;
-			}
-
-			analytics.ga.initialize();
-
+		recordPageView: makeSafeDebuggableGoogleAnalyticsFunc( function recordPageView(
+			urlPath,
+			pageTitle
+		) {
```
Comment (ockham) [inline client/lib/analytics/index.js]: I find it a bit confusing that the `function` is named. Have you considered using arrow notation?

File: client/lib/analytics/index.js (hunk 1 above)
Comment (ockham) [reply client/lib/analytics/index.js]: `makeAnalyticsGreatAgain`

File: client/lib/analytics/index.js (hunk 3 above)
Comment (sirreal) [reply client/lib/analytics/index.js]: That was intentional to make sure we have `func.name`, which makes the debugging output more informative:

https://github.com/Automattic/wp-calypso/blob/08d999c76af71b20b3d4c361594ef58156f25bd5/client/lib/analytics/index.js#L522-L525

That way we can write

```
[Disallowed] analytics recordEvent( /* args */ )
```

And you can inspect more closely _what_ GA event was fired with what args.

**Comment by sirreal** [comment]
😂  I need to start writing more buggy PRs, like this one!

https://github.com/Automattic/wp-calypso/pull/20119/commits/dcee18f4be867c5dac73249c8e00327005478826 fixes a bug where the wrapped function wasn't being passed the parameters correctly 😬 

This should have caused more concern (from description):

allowed

Fixed now:

fixed

PS: I never can remember whether `call` or `apply` is the array version 🤔 
PPS: I have not been in the right state of mind. Just `func( ...args )` 🤦‍♂️

**Comment by sirreal** [comment]
Just added a test for the wrapper that would have caught the bug.

Y'all think this can 🚢 ?

**Comment by tyxla** [comment]
🤦‍♂️ 

_makes a note to review less PRs on Friday_

LGTM, thanks for adding tests 👍

**Pull request merged**